farmerbot/internal/power_manager.go

```diff
@@ -0,0 +1,399 @@
+package internal
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+	"slices"
+	"sort"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog/log"
+	substrate "github.com/threefoldtech/tfchain/clients/tfchain-client-go"
+	zos "github.com/threefoldtech/zos/client"
+)
+
+// PowerManager manages the power of nodes
+type PowerManager struct {
+	state    *state
+	identity substrate.Identity
+}
+
+// NewPowerManager creates a new Power Manager
+func NewPowerManager(identity substrate.Identity, state *state) PowerManager {
+	return PowerManager{
+		state:    state,
+		identity: identity,
+	}
+}
+
+// PowerOn sets the node power state ON
+func (p *PowerManager) PowerOn(sub Sub, nodeID uint32) error {
+	log.Info().Uint32("node ID", nodeID).Str("manager", "power").Msg("POWER ON")
+	p.state.m.Lock()
+	defer p.state.m.Unlock()
+
+	node, ok := p.state.nodes[nodeID]
+	if !ok {
+		return fmt.Errorf("node %d is not found", nodeID)
+	}
+
+	if node.powerState == on || node.powerState == wakingUP {
+		return nil
+	}
+
+	_, err := sub.SetNodePowerTarget(p.identity, nodeID, true)
+	if err != nil {
+		return errors.Wrapf(err, "failed to set node %d power target to up", nodeID)
+	}
+
+	node.powerState = wakingUP
+	node.lastTimePowerStateChanged = time.Now()
+
+	return p.state.updateNode(node)
+}
+
+// PowerOff sets the node power state OFF
+func (p *PowerManager) PowerOff(sub Sub, nodeID uint32) error {
+	log.Info().Uint32("node ID", nodeID).Str("manager", "power").Msg("POWER OFF")
+	p.state.m.Lock()
+	defer p.state.m.Unlock()
+
+	node, ok := p.state.nodes[nodeID]
+	if !ok {
+		return fmt.Errorf("node %d is not found", nodeID)
+	}
+
+	if node.powerState == off || node.powerState == shuttingDown {
+		return nil
+	}
+
+	if node.neverShutDown {
+		return errors.Errorf("cannot power off node, node is configured to never be shutdown")
+	}
+
+	if node.PublicConfig.HasValue {
+		return errors.Errorf("cannot power off node, node has public config")
+	}
+
+	if node.timeoutClaimedResources.After(time.Now()) {
+		return errors.Errorf("cannot power off node, node has claimed resources")
+	}
+
+	onNodes := p.state.filterNodesPower([]powerState{on})
+
+	if len(onNodes) < 2 {
+		return errors.Errorf("cannot power off node, at least one node should be on in the farm")
+	}
+
+	_, err := sub.SetNodePowerTarget(p.identity, nodeID, false)
+	if err != nil {
+		return errors.Wrapf(err, "failed to set node %d power target to down", nodeID)
+	}
+
+	node.powerState = shuttingDown
+	node.lastTimePowerStateChanged = time.Now()
+
+	return p.state.updateNode(node)
+}
+
+// FindNode finds an available node in the farm
+func (p *PowerManager) FindNode(sub Sub, nodeOptions NodeOptions) (uint32, error) {
```

Comment:
It doesn't feel like a `FindNode` belongs to something called `power manager`

Note: I already stated multiple times that I don't like that we have that those called managers which all share the same state, so basically they can become small **parameterised** methods on the main farmerbot object.